c++/src/kj/compat/http.c++

```diff
@@ -6177,74 +6338,118 @@ private:
     }
 
-    kj::Own<kj::AsyncIoStream> accept(uint statusCode,
-                                      kj::StringPtr statusText,
-                                      const HttpHeaders& headers) override {
+    void accept(uint statusCode, kj::StringPtr statusText, const HttpHeaders& headers) override {
       KJ_REQUIRE(statusCode >= 200 && statusCode < 300, "the statusCode must be 2xx for accept");
-
-      auto headersCopy = kj::heap(headers.clone());
-
-      auto pipe = newTwoWayPipe();
-
-      kj::Own<kj::AsyncIoStream> io =
-          kj::heap<DelayedEofIoStream>(kj::mv(pipe.ends[0]),
-              task.attach(kj::addRef(*this), kj::mv(headersCopy)));
-
-      fulfiller->fulfill(HttpClient::ConnectResponse {
-        statusCode,
-        statusText,
-        headersCopy.get(),
-        kj::mv(io),
-      });
-      return kj::mv(pipe.ends[1]);
+      respond(statusCode, statusText, headers);
     }
 
     kj::Own<kj::AsyncOutputStream> reject(
         uint statusCode,
         kj::StringPtr statusText,
         const HttpHeaders& headers,
-        kj::Maybe<uint64_t> expectedBodySize) override {
+        kj::Maybe<uint64_t> expectedBodySize = nullptr) override {
       KJ_REQUIRE(statusCode < 200 || statusCode >= 300,
           "the statusCode must not be 2xx for reject.");
-      auto statusTextCopy = kj::str(statusText);
-      auto headersCopy = kj::heap(headers.clone());
+      auto pipe = kj::newOneWayPipe();
+      respond(statusCode, statusText, headers, kj::mv(pipe.in));
+      return kj::mv(pipe.out);
+    }
 
-      auto pipe = newOneWayPipe(nullptr);
+  private:
+    struct StreamsAndFulfiller {
+      // inner is the original owned stream that represents the bidirectional connection.
+      // guarded is the wrapped/guarded stream that wraps a reference to
+      // the first but blocks reads until the connection is accepted or rejected.
+      // This will be handed off when getConnectStream() is called.
+      // The fulfiller is used to resolve the guard for the second stream. This will
+      // be fulfilled or rejected when accept/reject is called.
+      kj::Own<kj::AsyncIoStream> inner;
+      kj::Own<kj::AsyncIoStream> guarded;
+      kj::Own<kj::PromiseFulfiller<void>> fulfiller;
+    };
 
-      auto wrapper = kj::heap<DelayedEofInputStream>(
-          kj::mv(pipe.in), task.attach(kj::addRef(*this)));
+    kj::Own<kj::PromiseFulfiller<HttpClient::ConnectRequest::Status>> fulfiller;
+    StreamsAndFulfiller streamAndFulfiller;
+    bool connectStreamDetached = false;
+
+    StreamsAndFulfiller initStreamsAndFulfiller(kj::Own<kj::AsyncIoStream> stream) {
+      auto paf = kj::newPromiseAndFulfiller<void>();
+      auto wrapped = kj::Own<kj::AsyncIoStream>(stream.get(), kj::NullDisposer::instance);
+      auto guarded = kj::heap<AsyncIoStreamWithGuards>(
+          kj::mv(wrapped),
+          kj::Maybe<HttpInputStreamImpl::ReleasedBuffer>(nullptr),
+          kj::mv(paf.promise)).attach(kj::addRef(*this));
```

Comment:
This code is called in the constructor of `ConnectResponseImpl` which means that if `ConnectResponseImpl` is allocated using `kj::refcounted` we end up with an error here because the refcounted didn't finish incrementing the ref count (or at least that's what I believe happens).

In any case I get a crash here because of that. Wonder why the tests didn't pick this up.

Reply:
Ugh... doh, that's right.  Not sure why it didn't pick it up. Very odd. Ok, I'll update this so that setting the stream happens in a separate call after allocating the instance.